src/libcollections/vec.rs

```diff
@@ -1799,95 +1826,92 @@ impl<T> Drop for IntoIter<T> {
     }
 }
 
-/// An iterator that drains a vector.
-#[unsafe_no_drop_flag]
-#[unstable(feature = "collections",
-           reason = "recently added as part of collections reform 2")]
-pub struct Drain<'a, T:'a> {
-    ptr: *const T,
-    end: *const T,
-    marker: PhantomData<&'a T>,
+/// A draining iterator for `Vec<T>`.
+#[unstable(feature = "collections", reason = "recently added")]
+pub struct Drain<'a, T: 'a>
+{
+    /// Index of tail to preserve
+    tail_start: usize,
+    /// Length of tail
+    tail_len: usize,
+    /// Current remaining range to remove
+    iter: slice::Iter<'a, T>,
+    vec: *mut Vec<T>,
+    _marker: PhantomData<&'a mut Vec<T>>,
```

Comment:
Ah that's why I was thinking of using `IterMut` instead of `Iter` (to take out the mutable borrow as opposed to a shared one). Although I'm not sure what effect that would have on this in terms of variance.